Some minor docker improvements

g2flyer opened the pull request:
A few minor improvements in docker code to
- docker & docker-compose consistent
  -  consistent tagging
  -  build only via docker to prevent double builds
  - consistent support of ADD_APT_PKGS
- some additional docker related make targets, e.g., additional stop targets, test without complete rebuild and _dev targets to setup development containers
- some minor dockerfile changes to be more consistent

BTW: in readme i've noticed we suggest in our examples to put user-name into container names. This seems a wise idea for multi-user settings. Should we also adopt it in the makefiles?  Easy to do but haven't changed it yet as i'm not sure if some other places count on the container names being what they are ...
Along the lines, the might include in name in container (and, more importantly, image) also SGX-mode (where also hw-mode in testing seems to have gone lost but see also PR#463, interpreter, ledger, ...?

cmickeyb commented on docker/pdo_client.dockerfile:
```diff
@@ -37,6 +37,9 @@ USER $UNAME
 
+ARG PDO_SOURCE_ROOT=/project/pdo/src
+ENV PDO_SOURCE_ROOT=${PDO_SOURCE_ROOT}
+
```
suggest we drop this. it should not be configurable.

g2flyer replied:
I didn't realize that one is supposed to run start_client which then all sets this (and below).  So definitely will revert and see whether maybe we can also do the start_client stuff automagically ..

cmickeyb replied:
Ideally... start_client.sh would be run kind of like bashrc. But there were a bunch of other usability issues. Open to suggestions.

cmickeyb commented on docker/pdo_client.dockerfile:
```diff
@@ -54,7 +57,7 @@ ENV PDO_LOG_LEVEL=${PDO_LOG_LEVEL}
 WORKDIR /project/pdo
-COPY --chown=${UNAME}:${UNAME} repository /project/pdo/src
+COPY --chown=${UNAME}:${UNAME} repository ${PDO_SOURCE_ROOT}
 
```
same. there is no reason to make pdo_source_root configurable.

g2flyer replied:
see aboe

cmickeyb commented on docker/pdo_client.dockerfile:
```diff
@@ -66,10 +69,13 @@ COPY --chown=${UNAME}:${UNAME} tools/*.sh ./
+
+ARG PDO_LEDGER_KEY_ROOT=/project/pdo/xfer/ccf/keys/
+ENV PDO_LEDGER_KEY_ROOT=${PDO_LEDGER_KEY_ROOT}
 
```
same. we control the layout of the container file system. we choose the structure.

g2flyer replied:
see above

cmickeyb reviewed (changes requested):
Overall comment: please be consistent with "{}" or "()" for variable expansion in the Makefile. I don't have a strong feeling either way, but I would strongly prefer to go with one form or the other.

g2flyer commented:
@cmickeyb Mic, i reworked the PR after our discussion, the PR description is now updated to the current changes (which essentially should address your comments as mentioned above)

bvavala commented on docker/pdo_client.dockerfile:
```diff
@@ -65,13 +65,13 @@ COPY --chown=${UNAME}:${UNAME} tools/*.sh ./
 
-ARG PDO_HOSTNAME
-ENV PDO_HOSTNAME=$PDO_HOSTNAME
+ARG PDO_HOSTNAME=localhost
+ENV PDO_HOSTNAME=${PDO_HOSTNAME}
 
-ARG PDO_LEDGER_URL
-ENV PDO_LEDGER_URL=$PDO_LEDGER_URL
```
Thinking about this, it seems appropriate not to include a ledger url in the image, but I am wondering if even the hostname should be provided at a later time.

g2flyer replied:
I see your point. I think a valid PDO_HOSTNAME var s required by some of our build scripts. That said, in that case we could (should) drop the ARG part but still retain the ENV so the build works?

cmickeyb replied:
I'm writing up a more complete version of what needs to happen with docker. will push in an issue for discussion. We (mea culpa) are still mixing up 1) a generic image that anyone could deploy on any network and 2) an image pre-configured for a specific network. I think we should make ALL images that we build preconfigured with hostname as localhost and ledger as 127.0.0.1:6600. Then use the start_* scripts to set what actually needs to be done. Keys are already provided along similar lines.

cmickeyb replied:
> I thoutht i could split check_pdo_env in runtime and build-time check but alas with some tests scripts during build it willl also be required. 

which scripts do you think we run at build time? specifically for docker...

bvavala commented on docker/pdo_services.dockerfile:
```diff
@@ -55,12 +55,9 @@ EXPOSE 7001 7002 7003 7004 7005
 EXPOSE 7201 7202 7203 7204 7205
 
-ARG PDO_HOSTNAME
+ARG PDO_HOSTNAME=localhost
 ENV PDO_HOSTNAME=$PDO_HOSTNAME
 
-ARG PDO_LEDGER_URL
-ENV PDO_LEDGER_URL=$PDO_LEDGER_URL
```
(same comment as above)

cmickeyb replied:
as i mentioned above... the only good reason for having these in the dockerfile is that they simplify construction of pre-configured images. but... i'm with you... pdo-hostname and pdo-ledger-url should NOT be build variables. we should pre-configure for localhost and actually set the values we want in the start_* scripts.

cmickeyb replied:
if you are keeping hostname then you have to keep the ledger_url. there is no reason whatsoever to have one without the other. either drop both or keep both.

bvavala commented on docker/services_base.yaml:
```diff
@@ -17,18 +17,9 @@ version: "3.4"
 services:
   services_container:
-    image: pdo_services
-    build:
```
+1 to remove this and consolidate the build
(the other way around of using just docker-compose worked as well)

cmickeyb replied:
again. I dont think this is a solution.
or maybe its a solution that we'll end up changing again in a couple weeks.

bvavala reviewed (commented):
Looks good overall, just one comment.

cmickeyb commented on docker/Makefile:
```diff
@@ -136,14 +141,14 @@ TEST_FILES += -f services_base.yaml
-test : clean_config clean_repository build_test
-	PDO_VERSION=${PDO_VERSION} docker-compose $(TEST_FILES) up --abort-on-container-exit
-	PDO_VERSION=${PDO_VERSION} docker-compose $(TEST_FILES) down
 
```
this is lovely in theory... but not particularly useful in practice. if i'm re-running the tests its because i made changes to the repository. so... does this really add more value than the complexity it adds?

g2flyer replied:
it tremendously helps, e.g., if you change anything related to docker building which is _not_ taking from the repository (which of course arguably could be seen as an issue)

cmickeyb replied:
but how often to you do this? that's a very small fraction of the time.

cmickeyb commented on docker/base.yaml:
```diff
@@ -18,9 +18,6 @@ version: "3.4"
-        PDO_VERSION: ${PDO_VERSION:-latest}
+    # build containers only via Makefile and directly via docker build
+    # as we use containers also outside of docker-compose to prevent
+    # double-build
```
ah... way ahead of me! see my comment about add SGX_MODE to the versions & i think this will help a LOT.

g2flyer replied:
see above, i don't think adding sgx-mode to tag is the right thing to do

cmickeyb commented on docker/pdo_ccf.dockerfile:
```diff
@@ -20,9 +20,9 @@ FROM pdo_ccf_base:${PDO_VERSION}
 # -----------------------------------------------------------------
-ARG REBUILD 0
+ARG REBUILD=0
 
```
ok.. for consistency i'm fine with this. but... turns out that "=" is optional.

g2flyer replied:
i know it's optional, just did it for consistency ...

cmickeyb commented on docker/pdo_ccf.dockerfile:
```diff
@@ -49,11 +49,8 @@ RUN /project/pdo/tools/build_ccf.sh
+ENV PDO_HOSTNAME=${PDO_HOSTNAME}
 
 # Note that the entry point when specified with exec syntax
 # can be extended through the docker run interface far more
```
this needs more discussion. adding the hostname and ledger url allows construction of pre-configured images (that is, the image contains the url). in practice, however, its WAY easier to just specify the interface and ledger in the startup script parameters.

g2flyer replied:
this is an outdated change regardless of github not marking it as such. In fact, the actual change is exactly the opposite (i.e., exactly what you want?)

cmickeyb commented on docker/pdo_client.dockerfile:
```diff
@@ -68,16 +65,13 @@ COPY --chown=${UNAME}:${UNAME} tools/*.sh ./
+NOTE: to setup environment, \"source /project/pdo/tools/start_client.sh\"\n\
+   with the arguments \"--mode build\" to build a new client environment (default mode)\n\
+   or \"--mode copy\" to copy one from the xfer directory"\
+' > /project/pdo/.bashrc
```
just create a .bashrc file & put this in the file. i will not approve changes that involve creating scripts in the dockerfile.

g2flyer replied:
yeah, i was thinking about that but then didn't really like to put a .bashrc file, so i decided on this.  But then in copy i can change file name, so while but it as pdo_client.bashrc  .and then COPY it ...

g2flyer replied:
fixed ...

cmickeyb commented on docker/tools/build_client.sh:
```diff
@@ -15,9 +15,9 @@
+# to get build without (ignored) errors
+export PDO_HOSTNAME=localhost
+export PDO_LEDGER_URL=https://127.0.0.1:6600
 
```
still suggest we pull these changes into a separate PR. we should not have to set these until we run.
also... push these into environment.sh. not here.

cmickeyb commented on docker/tools/build_services.sh:
```diff
@@ -13,9 +13,9 @@
-export PDO_LEDGER_URL=
+# to get build without (ignored) errors
+export PDO_HOSTNAME=localhost
+export PDO_LEDGER_URL=https://127.0.0.1:6600
```
same

g2flyer commented on bin/lib/common.sh:
```diff
@@ -82,10 +82,13 @@ function force_to_ip()
+    : "${PDO_LEDGER_TYPE:-$(die Missing environment variable PDO_LEDGER_TYPE)}"
+    : "${PDO_LEDGER_URL:-$(die Missing environment variable PDO_LEDGER_URL)}"
+    : "${PDO_SOURCE_ROOT:-$(die Missing environment variable PDO_SOURCE_ROOT)}"
+}
```
It actually does, as otherwise there were error (messages) during build ....

cmickeyb commented on bin/lib/common.sh:
```diff
@@ -47,7 +47,7 @@ function yell () {
 function die() {
-    recho "$(basename $0): $*" >&2
+    recho "$(basename $0): ERROR: $*" >&2
     exit 111
```
not to be pedantic... but did you check the use of this in the MANY other places its used? is "error:" really necessary?

g2flyer replied:
well,  we do an exit afterwards and so suggests the function name. So clearly it is considered an error (except of course the exit is ignored on some places where we do `: good-condition || $(die ....)` that this is clearly still semantically an error.
I actulaly also add a WARN to yell but we seem to use yell also for non-warning cases so i dropped that.  The main reason i added that is that i wanted to be able to grep for these messages in the voluminous log-fiels .  I guess there is a way to also grep for the color red but i think these seemed cleaner and works even on b/w terminals ...

g2flyer commented on docker/Makefile:
```diff
@@ -136,14 +141,14 @@ TEST_FILES += -f services_base.yaml
-		docker-compose $(TEST_FILES) build \
-			--build-arg PDO_VERSION=${PDO_VERSION}
+build_test : repository build_services build_ccf build_client
+
```
this is already done? There shouldn't be any build section in the compose-files anymore but we of course still needed the build targets in the makefile to build using docker build?

g2flyer commented on docker/Makefile:
```diff
@@ -60,36 +58,43 @@ all : $(addprefix build_,$(IMAGES))
 run_ccf_dev : build_ccf
-	@ docker run -v $(DOCKER_DIR)/xfer/:/project/pdo/xfer -v ${PDO_SOURCE_ROOT}/docker/tools/:/project/pdo/tools -v ${PDO_SOURCE_ROOT}/:/project/pdo/src --entrypoint=/bin/bash -it --network host --name ccf_container_dev -P pdo_ccf:${PDO_VERSION} -c "stty cols $$(tput cols) rows $$(tput lines) && bash"
+	@ docker run -v $(DOCKER_DIR)/xfer/:/project/pdo/xfer -v ${PDO_SOURCE_ROOT}/docker/tools/:/project/pdo/tools -v ${PDO_SOURCE_ROOT}/:/project/pdo/src --entrypoint=/bin/bash -it --network host --name ccf_container_dev -P pdo_ccf:$(PDO_VERSION) -c "stty cols $$(tput cols) rows $$(tput lines) && bash"
 
```
that is true once we go to a registry and unlike windows and mac seems docker on linux doesn't offer some convenience uid/gid mapping.  That said, isn't this also related what i mentioned above as question in the PS: whether we should follow also here the naming convention of including user into image names so we can also safely develop on multi-user machine?  In which case the _dev versions would also work fine.  Is of course also related to the below naming convention regarding sgx-mode and alike (and indirectly what our strategy would be regarding "production" setups and related to that how to manage/organise the regstiry?

cmickeyb replied:
suggest you move all the dev tags to a separate make file that can be included. and include LOTS of comments about the limitations of these targets. They are very useful... but the problems are not intuitive.

cmickeyb commented on docker/Makefile:
```diff
@@ -60,36 +58,43 @@ all : $(addprefix build_,$(IMAGES))
-		--tag pdo_$*:${PDO_VERSION} \
+		--build-arg PDO_VERSION=$(PDO_VERSION) \
+		--tag pdo_$*:$(PDO_VERSION) \
 		--file $(DOCKER_DIR)/pdo_$*.dockerfile .
```
naming might be easier with tags, but i agree that the best way to do this is to push it into the container name itself. good suggestion!

cmickeyb commented on docker/pdo_client.bashrc:
```diff
@@ -0,0 +1,5 @@
+NOTE: to setup environment, \"source /project/pdo/tools/start_client.sh\"\n\
+   with the arguments \"--mode build\" to build a new client environment (default mode)\n\
+   or \"--mode copy\" to copy one from the xfer directory"
+
```
put this in the tools directory.

g2flyer replied:
i wouldn't have seen this really as a tool and we still have to do dockerfile-action as the copying of tools is not enough?  But if you prefer i can put it there and then replace the current COPY with with RUN 'ln -s ...' command ...

g2flyer commented:
@cmickeyb the latest version should address your comments and merge in the changes from your branch, with slight modifications for make.dev ...

bvavala reviewed (approved):
looks good
